code/drasil-lang/Language/Drasil/Chunk/Eq.hs

```diff
@@ -20,125 +27,173 @@ import Language.Drasil.Chunk.DefinedQuantity (DefinedQuantityDict, dqd, dqd')
 import Language.Drasil.Chunk.NamedIdea (mkIdea, nw)
 
-import Language.Drasil.ModelExpr.Math (defines)
-import Language.Drasil.Expr (Expr(FCall, C))
-import Language.Drasil.Expr.Math (sy)
+import Language.Drasil.ModelExpr (defines)
+import Language.Drasil.ModelExpr.Lang (ModelExpr)
+import Language.Drasil.Expr.Lang (Expr(FCall, C))
+import Language.Drasil.Expr (sy)
 import Language.Drasil.NounPhrase.Core (NP)
 import Language.Drasil.Space (mkFunction, Space)
 import Language.Drasil.Sentence (Sentence(EmptyS))
 import Language.Drasil.Stages (Stage)
 import Language.Drasil.Symbol (Symbol)
 import Language.Drasil.UID (UID)
 
--- | A QDefinition is a 'QuantityDict' with a defining expression ('Expr'), a definition ('Sentence'), and a domain (['UID']).
--- This chunk now contains almost enough information to generate code, definitions, and models.
--- Used for high-level quantities with equations and inputs.
-data QDefinition = EC
-  { _qua    :: DefinedQuantityDict
-  , _inputs :: [UID]
-  , _equat  :: Expr
-  }
-makeLenses ''QDefinition
-
--- | Finds the 'UID' of the 'QuantityDict' used to make the 'QDefinition'.
-instance HasUID        QDefinition where uid = qua . uid
--- | Finds the term ('NP') of the 'QuantityDict' used to make the 'QDefinition'.
-instance NamedIdea     QDefinition where term = qua . term
--- | Finds the idea contained in the 'QuantityDict' used to make the 'QDefinition'.
-instance Idea          QDefinition where getA = getA . (^. qua)
--- | Finds the 'Space' of the 'QuantityDict' used to make the 'QDefinition'.
-instance HasSpace      QDefinition where typ = qua . typ
--- | Finds the 'Symbol' of the 'QuantityDict' used to make the 'QDefinition'.
-instance HasSymbol     QDefinition where symbol = symbol . (^. qua)
--- | Finds the definition of 'QDefinition'.
-instance Definition    QDefinition where defn = qua . defn
--- | 'QDefinition's have a 'Quantity'.
-instance Quantity      QDefinition where
--- | Finds the defining expression of 'QDefinition'.
-instance DefiningExpr  QDefinition where defnExpr = equat
--- | Equal if 'UID's are equal.
-instance Eq            QDefinition where a == b = (a ^. uid) == (b ^. uid)
--- | Finds the units of the 'QuantityDict' used to make the 'QDefinition'.
-instance MayHaveUnit   QDefinition where getUnit = getUnit . view qua
--- | Displays the relation given by the expression in 'QDefinition'.
-instance Express       QDefinition where
-  express q = f (q ^. defnExpr)
+data QDefinition e where
+  QD :: Express e => DefinedQuantityDict -> [UID] -> e -> QDefinition e
+
+qdQua :: Express e => Lens' (QDefinition e) DefinedQuantityDict
+qdQua = lens (\(QD qua _ _) -> qua) (\(QD _ ins e) qua' -> QD qua' ins e)
+
+qdInputs :: Express e => Lens' (QDefinition e) [UID]
+qdInputs = lens (\(QD _ ins _) -> ins) (\(QD qua _ e) ins' -> QD qua ins' e)
+
+qdExpr :: Express e => Lens' (QDefinition e) e
+qdExpr = lens (\(QD _ _ e) -> e) (\(QD qua ins _) e' -> QD qua ins e')
```

Comment:
Unfortunately, the DataTypeContexts language extension is deprecated (and apparently does nothing anyways), so I switched to using GADTs here + DataDefinitions. `makeLenses` doesn't appear to work well with GADTs, so we manually create Lenses for them.

Reply:
I originally wanted to, at the type-level, only allow Exprs/ModelExprs as the "RHS" of QDefinitions with minimal code reuse, but now I'm starting to doubt the need for the GADT syntax + Express constraint on the constructor. Perhaps we should revert it back to what it was originally?

Reply:
You might want to investigate going with
```haskell
data Exprs = Simple | Model

data QDefinition (e :: 'Exprs) = EC
  { _qua    :: DefinedQuantityDict
  , _inputs :: [UID]
  , _equat  :: e
  }
```
where I've used "promotion" to promote a data type to a kind. This limits `e` to be of kind `Simple` or `Model` only.  https://downloads.haskell.org/~ghc/7.8.4/docs/html/users_guide/promotion.html 

I'm pretty sure my code isn't quite right, but something like that should work.

Another option is to use GADTs sort of directly, as follows:
```haskell
data QDefinition e where
  QDE :: DefinedQuantityDict -> [UID] -> Expr -> QDefinition Simple
  QDM :: DefinedQuantityDict -> [UID] -> ModelExpr -> QDefinition Model
```
which requires fewer extensions and fancy stuff, so it likely preferable.

Reply:
Thank you!

The first solution is very interesting. I'll try to play around with it in a small test area. The second solution also looks good, it's definitely an improvement over the constraints method!

Reply:
Just a small progress report.

I couldn't quite get the first solution working without GADTs, which then ends up looking like the second solution.

With the second solution, would we need to duplicate each of the smart constructors of the qdefinitions (to create simple and model constructors)? I don't seem to see an easy way to parameterize over the different Exprs nicely. Additionally, with #2783 in mind (assuming it's a good idea), we would potentially have `QDefinition Constant`s appearing, meaning that `QDefinitions` would have 3 kinds of smart constructors (1 for each language).

Reply:
What do you think of creating an empty typeclass (like the `Quantity` one) that is constrained to `Express`ible things for which we only define for `ModelExpr`, and `Expr`? With this, we could use less `Express` constraints everywhere and have more of this one, and since we would only instantiate it for `Expr` and `ModelExpr`, we have essentially the same effect as `Exprs`.